src/config/jobconfig.ts

```diff
@@ -19,172 +19,236 @@ import { ApiProperty } from "@nestjs/swagger";
+   * @param data JSON
+   * @returns
+   */
+  static parse(data: Record<string, any>) {
+    const type = data["jobType"];
+    const create = JobOperation.parse<CreateJobDto>(
+      createActions,
+      data["create"],
+    );
+    const read = undefined; //"read" in data ? oneOrMore(data["read"]).map((json) => parseReadAction(json["action"])) : [];
+    const update = JobOperation.parse<UpdateJobStatusDto>(
+      updateActions,
+      data["update"],
+    );
+    return new JobConfig(type, create, read, update);
+  }
+}
+
+export class JobOperation<DtoType> {
+  auth: Record<string, any> | undefined; //TODO replace with actual auth schema
+  actions: JobAction<DtoType>[];
+
+  constructor(
+    actions: JobAction<DtoType>[] = [],
+    auth: Record<string, any> | undefined,
+  ) {
+    this.actions = actions;
+    this.auth = auth;
+  }
+
+  static parse<DtoType>(
+    actionList: Record<string, JobActionClass<DtoType>>,
+    data: Record<string, any>,
+  ): JobOperation<DtoType> {
+    const auth = data["auth"];
+    const actionsData: any[] = oneOrMore(data["actions"] || []);
+    const actions = actionsData.map((json) =>
+      parseAction<DtoType>(actionList, json),
+    );
+    return new JobOperation<DtoType>(actions, auth);
+  }
 }
 
-function oneOrMore<T>(x: T[]|T): T[]  {
-    return Array.isArray(x) ? x : [x];
+/**
+ * Given a JSON object configuring a JobConfigAction.
+ *
+ * This is dispatched to registered constructors (see registerCreateAction) based on
+ * the "actionType" field of data. Other parameters are action-specific.
+ * @param data JSON configuration data
+ * @returns
+ */
+function parseAction<DtoType>(
+  actionList: Record<string, JobActionClass<DtoType>>,
+  data: Record<string, any>,
+): JobAction<DtoType> {
+  if (!("actionType" in data))
+    throw SyntaxError(`No action.actionType in ${JSON.stringify(data)}`);
+
+  const type = data.actionType;
+  if (!(type in actionList)) {
+    throw SyntaxError(`No handler found for actions of type ${type}`);
+  }
+
+  const actionClass = actionList[type];
+  return new actionClass(data);
 }
 
 /**
  * Superclass for all responses to Job changes
  */
 export interface JobAction<DtoType> {
-    // TODO should this be static? how to type that?
-    /**
-     * Action type, eg "url".
-     * 
-     * Not to be confused with JobConfig.type
-     */
-    //readonly actionType: string;
-    /**
-     * Validate the DTO, throwing an HttpException for problems
-     */
-    validate: (dto: DtoType) => Promise<void>;
-    /**
-     * Respond to the action
-     */
-    performJob: (job: JobClass) => Promise<JobClass>;
+  /**
+   * Validate the DTO, throwing an HttpException for problems
+   */
+  validate: (dto: DtoType) => Promise<void>;
+  /**
+   * Respond to the action
+   */
+  performJob: (job: JobClass) => Promise<JobClass>;
+  /**
+   * Return the actionType for this action. This should match the class's
```

Comment:
The static `JobAction.actionType` is used in parsing the config file. Method `jobAction.getActionType()` is used when executing the action (eg in error logs).
The type system doesn't enforce that they match, but hopefully this comment is enough.

Reply:
I'm confused on this comment.
What do you mean by `the system does not enforce that they match`? Do you mean JobAction.actionType and the value returned by JobAction.GetActionType()?